### PR TITLE
Add FileLu cloud storage backend

### PR DESCRIPTION
**Pull request by filelucom**
This pull request introduces the `FileLu` cloud storage backend to Rclone, allowing users to integrate their FileLu cloud storage with Rclone for seamless file list files, list folders, copy, move, uploads, downloads, sync, and management.

#### Checklist

- [x ] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ x] I have added tests for all changes in this PR if appropriate.
- [x ] I have added documentation for the changes if appropriate.
- [x ] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by ncw** [comment]
I had a brief look at the code - it looks good :-) I'll do a full review later in the process.

First what I want to know is how are [the integration tests going](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend)? Are you at the stage `test_all` works properly?

Next if you look through the docs above you'll see that you need to write some docs next.

I take it from your user name that you work for FileLu ?

Thank you for the contribution

**Comment by filelucom** [comment]
> I had a brief look at the code - it looks good :-) I'll do a full review later in the process.
> 
> First what I want to know is how are [the integration tests going](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend)? Are you at the stage `test_all` works properly?
> 
> Next if you look through the docs above you'll see that you need to write some docs next.
> 
> I take it from your user name that you work for FileLu ?
> 
> Thank you for the contribution

Thank you!
This is our first version of Rclone, so it's not perfect and still has room for improvement, but we will try our best. Yes, most of the commands are working. We still need to improve the mount command. I just added the filelu.md documentation.
Yes, I work for FileLu.
We appreciate all the feedback. Your feedback is incredibly valuable to us.

**Comment by ncw** [comment]
> Thank you! This is our first version of Rclone, so it's not perfect and still has room for improvement, but we will try our best. Yes, most of the commands are working.

Great

> We still need to improve the mount command.

If you get the integration tests working properly, then mount will work properly too.

> I just added the filelu.md documentation. Yes, I work for FileLu. We appreciate all the feedback. Your feedback is incredibly valuable to us.

Let me know when you've got the integration tests passing (or mostly passing) and I'll do a full review.

**Comment by kingston125** [comment]
> Let me know when you've got the integration tests passing (or mostly passing) and I'll do a full review.

 All the integration tests have passed. Thank you.

**Comment by ncw** [comment]
> All the integration tests have passed. Thank you.

Good news :-)

I don't see a change to [config.yaml](https://github.com/rclone/rclone/blob/master/fstest/test_all/config.yaml) - did you try the full integration test suite as here in the docs: https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#integration-tests

Thanks

**Comment by kingston125** [comment]
> > All the integration tests have passed. Thank you.
> 
> Good news :-)
> 
> I don't see a change to [config.yaml](https://github.com/rclone/rclone/blob/master/fstest/test_all/config.yaml) - did you try the full integration test suite as here in the docs: https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#integration-tests
> 
> Thanks

Hi,

Sorry for the delayed reply. I was busy rewriting the entire code, and during testing, it failed. I'm not sure how to fix it, but it builds successfully and works fine with most commands.

**Comment by ncw** [comment]
How is the testing going @kingston125 ?

**Comment by kingston125** [comment]
> How is the testing going @kingston125 ?

It still fails when running the test. If you can, please help me check and fix it so that it passes the test. I would really appreciate it. It successfully builds and can run on any platform, and it works, but I’m not sure why it still doesn’t pass internal testing.

Thank you.

**Comment by kingston125** [comment]
Hi @ncw 
Does the backend need to pass all tests before being merged into Rclone? All checks have passed, but testing is still facing some issues. However, the backend and Rclone commands are working fine on all platforms.

Thank you.

**Comment by ncw** [comment]
Hi @kingston125 - sorry for the delay responding. Ideally we'd get as many tests passing as possible and ones which don't pass we mark as not passing in `fstest/test_all/config.yaml`.

How are you doing with this - is this ready to merge for v1.70?

**Comment by kingston125** [comment]
> How are you doing with this - is this ready to merge for v1.70?

Yes, this is ready to merge for v1.70.
Thank you Nick.

**Comment by ncw** [comment]
Hi @kingston125 

I've been through the code and fixed up some things - you'll see I've squashed your commits and added another commit on top. Can you pull back the changed code and work on top of that please.

The integration tests are some way from passing. You can run this truncated set with `go test -v -short` - these should really all pass but we are getting these failures

```
        --- FAIL: TestIntegration/FsMkdir/FsMkdirRmdirSubdir (8.93s)
        --- FAIL: TestIntegration/FsMkdir/FsListDirNotFound (0.32s)
        --- FAIL: TestIntegration/FsMkdir/FsPutZeroLength (1.33s)
        --- FAIL: TestIntegration/FsMkdir/FsPutFiles (10.14s)
        --- FAIL: TestIntegration/FsMkdir/FsUploadUnknownSize (8.01s)
            --- FAIL: TestIntegration/FsMkdir/FsUploadUnknownSize/FsUpdateUnknownSize (5.33s)
```

Can you get those ones to pass?

Other comments on the code:

1. This is missing calls to the pacer. If you look at an example backend you'll see api calls are wrapped in a pacer.  Without this pacer, rclone will not retry network errors or obey `--low-level-retries` so it is really important. https://github.com/rclone/rclone/blob/205667143c807103b493e94bc0a708e10c435598/backend/box/box.go#L408-L411
2. This is not using rclone's HTTP stack. Can you make it use `lib/rest` please? Without that features like `--bwlimit`, `--dump headers`, `--tpslimit`, `--conntimeout`, `--timeout` etc won't work. At minimum it needs to use rclone's http transport but I'd prefer if it used `lib/rest` like the example backend (eg box).
3. The backend can clearly do DirMove, Copy according to the backend commands but these aren't implemented as optional features.
4. This has a lot of debugs in which will need to be commented out at some point!

So to merge we need the integration tests to pass and ideally 1 and 2 from above. The optional features can wait.

Thank you.

**Comment by kingston125** [comment]
Thank you for the update and information. I already refactored the code to to use pacer.Call. For the FAIL test, when I run the functions ls, mkdir, and rmdir.. etc, everything works and functions correctly, but I don’t understand why it fails when running the test. Is it possible to skip the test?

**Comment by ncw** [comment]
> Thank you for the update and information. I already refactored the code to to use pacer.Call. 

Great

> For the FAIL test, when I run the functions ls, mkdir, and rmdir.. etc, everything works and functions correctly, but I don’t understand why it fails when running the test. Is it possible to skip the test?

The tests test the edge cases which don't happen in normal operation. Getting them all to pass will make sure all the rclone features work properly.

**Comment by kingston125** [comment]
> The tests test the edge cases which don't happen in normal operation. Getting them all to pass will make sure all the rclone features work properly.

Thank you for the info. My team is currently working on passing those tests. Out of the six tests you shared above, we've already cleared four, so we're almost there with just two remaining. We continue working on it.
For #3 The backend can clearly do DirMove, Copy according to the backend commands but these aren't implemented as optional features.
Could you please clarify what you mean by "implemented as optional features"? We understand that the backend can perform DirMove, Copy, and similar operations via commands, but what does it mean to implement them as optional features, and how should we go about doing that?

Thank you.

**Comment by kingston125** [comment]
> > > The tests test the edge cases which don't happen in normal operation. Getting them all to pass will make sure all the rclone features work properly.
> 
> > 
> 
> > Thank you for the info. My team is currently working on passing those tests. Out of the six tests you shared above, we've already cleared four, so we're almost there with just two remaining.
> 
> 
> 
> Great - how are you getting on with that?
> 
> 
> 
> > For #3 The backend can clearly do DirMove, Copy according to the backend commands but these aren't implemented as optional features. Could you please clarify what you mean by "implemented as optional features"? We understand that the backend can perform DirMove, Copy, and similar operations via commands, but what does it mean to implement them as optional features, and how should we go about doing that?
> 
> 
> 
> The optional features are described here
> 
> 
> 
> https://github.com/rclone/rclone/blob/1f8373fae8a2a059aa5a0d5d145ef434ca3feb3e/fs/features.go#L516-L556
> 
> 
> 
> You create them in your backend (eg box)
> 
> 
> 
> https://github.com/rclone/rclone/blob/1f8373fae8a2a059aa5a0d5d145ef434ca3feb3e/backend/box/box.go#L937-L946
> 
> 
> 
> And assert that you have implemented them
> 
> 
> 
> https://github.com/rclone/rclone/blob/1f8373fae8a2a059aa5a0d5d145ef434ca3feb3e/backend/box/box.go#L1737
> 
> 
> 
> If you implement them like that then the normal `rclone move`, `rclone copy` will be able to do server side moves and copies and other parts of rclone (like `rclone mount`) will use them for renaming files, etc.
> 
> 
> 
> Does that make sense?

Hi,
Everything is working, mount and sever site move, copy... are supported.

**Comment by kingston125** [comment]
I run the short test and they all passed.

File: backend/filelu/filelu.go [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+
+// Purge deletes the directory and all its contents
+func (f *Fs) Purge(ctx context.Context, dir string) error {
+
+	fullPath := path.Join(f.root, dir)
+	if fullPath != "" {
+		fullPath = "/" + strings.Trim(fullPath, "/")
+	}
+
+	// Step 1: Check if folder is empty
```
Comment (ncw) [inline backend/filelu/filelu.go:342]: Purge should remove a directory and **all** of its contents. So don't check if it is empty. If your backend can't do that then remove the Purge method.

File: backend/filelu/filelu.go [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+type Fs struct {
+	name       string
+	root       string
+	opt        Options
+	features   *fs.Features
+	endpoint   string
+	pacer      *pacer.Pacer
+	srv        *rest.Client
+	client     *http.Client
+	isFile     bool
```
Comment (ncw) [inline backend/filelu/filelu.go:80]: Your backend shouldn't need to know it is pointing at a file so remove this and its users. You return `fs.ErrorIsFile` if you detect it is a file but the Fs should reference the parent of the file so it behaves perfectly normally otherwise looking at the file parent directory.

File: backend/filelu/filelu.go [hunk 3]
```diff
@@ -0,0 +1,756 @@
...
+			size:    file.Size,
+			modTime: time.Now(),
+		}
+		entries = append(entries, obj)
+	}
+	return entries, nil
+}
+
+// ListR lists the objects and directories of the Fs recursively
+func (f *Fs) ListR(ctx context.Context, dir string, callback fs.ListRCallback) error {
```
Comment (ncw) [inline backend/filelu/filelu.go:423]: If your backend can't natively do recursive listings then don't implement this. Rclone has a much more sophisticated walkDir built in which it will use if you don't implement `ListR`

File: backend/filelu/filelu_client.go [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+	"net/url"
+	"strings"
+
+	"github.com/rclone/rclone/backend/filelu/api"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/fserrors"
+	"github.com/rclone/rclone/lib/rest"
+)
+
+// CreateFolderResponse represents the response for creating a folder.
```
Comment (ncw) [inline backend/filelu/filelu_client.go:19]: These type definitions for the API should go in api/types.go

File: backend/filelu/filelu_object.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+		o.fs.endpoint, url.QueryEscape(fileCode), url.QueryEscape(o.fs.opt.Key))
+
+	var result struct {
+		Status int    `json:"status"`
+		Msg    string `json:"msg"`
+		Result []struct {
+			Hash string `json:"hash"`
+		} `json:"result"`
+	}
+	err := o.fs.pacer.Call(func() (bool, error) {
```
Comment (ncw) [inline backend/filelu/filelu_object.go:205]: If looking up a Hash takes an HTTP call then it should be marked as Slow in the Features struct `SlowHash: true`

**Review by ncw — CHANGES_REQUESTED** [review]
I've given this a more thorough review - please see comments inline.

Also

- Please make any methods on Fs which aren't part of the public interface start with a lower case letter so they aren't exported.
- Check all HTTP transactions are wrapped in Fs.pacer.CallXXX

I ran the entire test suite with

```
go install ./...
test_all -remotes TestFileLu:
```

And this did quite well

image

The main problems seem to be the encodings - they aren't being applied everywhere they should be would be my guess. If you can get them working then lots more things will work properly.

Test these (in the backend/filelu directory) with

   go test -v -remote TestFileLu: -run '^TestIntegration$/^FsMkdir$/^FsEncoding$/'

Thank you

**Comment by kingston125** [comment]
Thank you for the info. I will start working on it and will let you know when it’s completed..

File: backend/filelu/filelu.go [hunk 6]
```diff
@@ -0,0 +1,756 @@
...
+		size:    src.Size(),
+		modTime: src.ModTime(ctx),
+	}
+	fs.Infof(f, "Put: Successfully uploaded new file %q", src.Remote())
+	return newObject, nil
+}
+
+// Move the objects and directories
+func (f *Fs) Move(ctx context.Context, src fs.Object, remote string) (fs.Object, error) {
+	// Check if the source is a directory
```
[the remark this answers is not in the document]
Comment (kingston125) [reply backend/filelu/filelu.go:565]: Done that.

File: backend/filelu/filelu.go [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+		fs:      f,
+		remote:  destinationPath,
+		size:    src.Size(),
+		modTime: src.ModTime(ctx),
+	}, nil
+}
+
+// MoveToLocal moves the file or folder to the local file system.
+// It implements the fs.Fs interface and performs the move operation locally.
+func (f *Fs) MoveToLocal(ctx context.Context, remote string, localPath string) error {
```
[the remark this answers is not in the document]
Comment (kingston125) [reply backend/filelu/filelu.go:679]: Done that.

File: backend/filelu/filelu.go [hunk 8]
```diff
@@ -0,0 +1,756 @@
...
+	err = json.NewDecoder(resp.Body).Decode(&result)
+	if err != nil {
+		return fmt.Errorf("error decoding response: %w", err)
+	}
+
+	if result.Status != 200 {
+		return fmt.Errorf("error: %s", result.Msg)
+	}
+
+	fs.Infof(f, "Removed directory %q successfully", dir)
```
[the remark this answers is not in the document]
Comment (kingston125) [reply backend/filelu/filelu.go:330]: Done that.

File: backend/filelu/filelu.go [hunk 9]
```diff
@@ -0,0 +1,756 @@
...
+
+// Mkdir to create directory on remote server.
+func (f *Fs) Mkdir(ctx context.Context, dir string) error {
+	fullPath := path.Clean(f.root + "/" + dir)
+	_, err := f.createFolder(ctx, fullPath)
+	return err
+}
+
+// DeleteFile sends an API request to remove a file from FileLu
+func (f *Fs) DeleteFile(ctx context.Context, filePath string) error {
```
[the remark this answers is not in the document]
Comment (kingston125) [reply backend/filelu/filelu.go:160]: Done that.

File: backend/filelu/filelu.go (hunk 1 above)
Comment (kingston125) [reply backend/filelu/filelu.go:342]: Done that.

File: backend/filelu/filelu.go (hunk 2 above)
Comment (kingston125) [reply backend/filelu/filelu.go:80]: Done that.

File: backend/filelu/filelu.go (hunk 3 above)
Comment (kingston125) [reply backend/filelu/filelu.go:423]: Done that.

File: backend/filelu/filelu_client.go (hunk 4 above)
Comment (kingston125) [reply backend/filelu/filelu_client.go:19]: Done that.

File: backend/filelu/filelu_object.go (hunk 5 above)
Comment (kingston125) [reply backend/filelu/filelu_object.go:205]: Done that.

**Comment by kingston125** [comment]
> go test -v -remote TestFileLu: -run '^TestIntegration$/^FsMkdir$/^FsEncoding$/'

The encoding test now passes. The code has also been updated as requested. Please let me know if I missed anything.

**Comment by kingston125** [comment]
For the DirMove, I tested it via mount and was able to rename and move directories as expected.

**Comment by ncw** [comment]
That is great - looking better :-)

image

Can you get the remaining tests to pass?

These were run with

```
go install ./...
test_all -remotes TestFileLu:
```

The failing tests are

```
cd fs/sync
go test -v -remote TestFileLu: -run '^(TestMoveWithDeleteEmptySrcDirs|TestMoveWithoutDeleteEmptySrcDirs)$'

cd vfs
go test -v -remote TestFileLu: -run '^TestFileSetModTime$/^cache=off,open=true,write=false$|^TestRWFileHandleWriteNoWrite$'" 
```

Can you make those pass, then run the entire test suite again and get everything to pass.

Thank you and well done for getting this far - nearly there :-)

**Comment by kingston125** [comment]
Hi,

cd fs/sync
go test -v -remote TestFileLu: -run '^(TestMoveWithDeleteEmptySrcDirs|TestMoveWithoutDeleteEmptySrcDirs)$'

cd vfs
go test -v -remote TestFileLu: -run '^TestFileSetModTime$/^cache=off,open=true,write=false$|^TestRWFileHandleWriteNoWrite$'" 

Is these two tests final?  I’m a little bit confused. For the TestFileSetModTime, is it mandatory? What if my backend doesn’t support it?

**Comment by ncw** [comment]
Run the tests and see if you can work out what is going on. It's ok for your backend not to support things - maybe it is not returning the correct error code. I this you've got the backend Features right since so many tests pass, but the problem could be there too.

**Comment by kingston125** [comment]
Hi,
Thank you for the info. I have updated the backend and config.yaml to skip some tests. Everything passed when I ran:
go test -v -short ./backend/filelu

The test below is now Passes:
cd fs/sync
go test -v -remote TestFileLu: -run '^(TestMoveWithDeleteEmptySrcDirs|TestMoveWithoutDeleteEmptySrcDirs)$'

The test below was added to config.yaml to ignore:
cd vfs
go test -v -remote TestFileLu: -run '^TestFileSetModTime$/^cache=off,open=true,write=false$|^TestRWFileHandleWriteNoWrite$'"

The test seems very strange, when I run `go test -v ./backend/filelu -remote remote:` it shown some test FAIL e.g 

           

>  --- FAIL: TestIntegration/FsMkdir/FsPutFiles (241.94s)
>             --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsListDirFile2 (1.58s)
>             --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsListR (1.61s)
>         
>            --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsListLevel2 (0.73s)
>         
>             --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsListFile1 (11.80s)
>           
>             --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsNewObjectCaseInsensitive (0.00s)
>             --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsListFile1and2 (11.87s)
>              --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsPurge (27.37s)
>             --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsPurgeRoot (27.63s)
>             
>             --- FAIL: TestIntegration/FsMkdir/FsPutFiles/FsMove (65.41s)
> 

But when Im run it one by one then it PASS. 
`go test -v -run 'TestIntegration/FsMkdir/FsPutFiles/FsListLevel2' ./backend/filelu -verbose`

=== NAME  TestIntegration/FsMkdir
    fstests.go:2692: Warning: this should produce fs.ErrorDirNotFound
--- PASS: TestIntegration (11.14s)
    --- PASS: TestIntegration/FsMkdir (10.22s)
        --- PASS: TestIntegration/FsMkdir/FsPutFiles (8.25s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListLevel2 (0.63s)
PASS
ok      github.com/rclone/rclone/backend/filelu 11.151s

                
Thank you.

**Comment by ncw** [comment]
In your config.yaml there are too many tests failing

Can you fix these please? You should be able to fix all the `FsEncoding` tests using `Enc` you have defined

```
	Enc encoder.MultiEncoder `config:"encoding"`
```

You need to use it like this (examples from the box backend)

```
github.com/rclone/rclone/backend/box/box.go
 612: 		Name: f.opt.Enc.FromStandardName(leaf),
 684: 			item.Name = f.opt.Enc.ToStandardName(item.Name)
 777: 		Name: f.opt.Enc.FromStandardName(leaf),
 996: 		Name: f.opt.Enc.FromStandardName(leaf),
1035: 		Name: f.opt.Enc.FromStandardName(leaf),
1326: 	name := f.opt.Enc.ToStandardName(childName)
1654: 		Name:              o.fs.opt.Enc.FromStandardName(leaf),
github.com/rclone/rclone/backend/box/upload.go
42: 		request.FileName = o.fs.opt.Enc.FromStandardName(leaf)
```

You should be able to get rid of all the encoding errors.

You will probably find some of the other tests start working at that point too.

The `FsPutFiles` failures are quite important. They probably indicate something is wrong with your listing routine.

Try to fix the **first** failing test by running it on its own, and when you get that working it will fix a lot of the other ones.

Once you've fixed those, please re-run the full integration test and report back.

Thank you

```yaml
    # FsEncoding tests
    - TestIntegration/FsMkdir/FsEncoding/control_chars
    - TestIntegration/FsMkdir/FsEncoding/dot
    - TestIntegration/FsMkdir/FsEncoding/dot_dot
    - TestIntegration/FsMkdir/FsEncoding/punctuation
    - TestIntegration/FsMkdir/FsEncoding/leading_space
    - TestIntegration/FsMkdir/FsEncoding/leading_tilde
    - TestIntegration/FsMkdir/FsEncoding/leading_CR
    - TestIntegration/FsMkdir/FsEncoding/leading_LF
    - TestIntegration/FsMkdir/FsEncoding/leading_HT
    - TestIntegration/FsMkdir/FsEncoding/leading_VT
    - TestIntegration/FsMkdir/FsEncoding/leading_dot
    - TestIntegration/FsMkdir/FsEncoding/trailing_space
    - TestIntegration/FsMkdir/FsEncoding/trailing_CR
    - TestIntegration/FsMkdir/FsEncoding/trailing_LF
    - TestIntegration/FsMkdir/FsEncoding/trailing_HT
    - TestIntegration/FsMkdir/FsEncoding/trailing_VT
    - TestIntegration/FsMkdir/FsEncoding/trailing_dot
    - TestIntegration/FsMkdir/FsEncoding/URL_encoding
    # FsPutFiles failures
    - TestIntegration/FsMkdir/FsPutFiles/FsListDirFile2
    - TestIntegration/FsMkdir/FsPutFiles/FsListR
    - TestIntegration/FsMkdir/FsPutFiles/FsListLevel2
    - TestIntegration/FsMkdir/FsPutFiles/FsListFile1
    - TestIntegration/FsMkdir/FsPutFiles/FsListFile1and2
    - TestIntegration/FsMkdir/FsPutFiles/FsPurge
    - TestIntegration/FsMkdir/FsPutFiles/FsPurgeRoot
    - TestIntegration/FsMkdir/FsPutFiles/FsMove
    - TestIntegration/FsMkdir/FsPutFiles/FromRoot
    - TestIntegration/FsMkdir/FsPutFiles/FromRoot/List
    - TestIntegration/FsMkdir/FsPutFiles/FromRoot/ListEntries
    - TestIntegration/FsMkdir/FsPutFiles/FromRoot/Put
    - TestIntegration/FsMkdir/FsPutFiles/FromRoot/Put/Remove
    - TestIntegration/FsMkdir/FsPutFiles/ObjectRemove
    # fs/operations
    - TestCheckSum
    - TestCheckSumDownload
    - TestHashSums/Md5
    - TestReadFile
    - TestCopyURL
    - TestMoveFileWithIgnoreExisting
    #vfs
    - TestFileSetModTime/cache=off,open=false,write=false
    - TestFileSetModTime/cache=off,open=true,write=false
    - TestRWFileHandleWriteNoWrite

```

**Comment by kingston125** [comment]
> Can you fix these please? You should be able to fix all the `FsEncoding` tests using `Enc` you have defined
I am so confused when I run: 
`go test -v -run 'TestIntegration/FsMkdir/FsEncoding/' ./backend/filelu -verbose `
it will all FAIL,
But then I run each FAIL one by one e.g:
`go test -v -run 'TestIntegration/FsMkdir/FsEncoding/dot' ./backend/filelu -verbose `

then it shows PASS. 
=== NAME TestIntegration/FsMkdir
fstests.go:2692: Warning: this should produce fs.ErrorDirNotFound
--- PASS: TestIntegration (24.36s)
--- PASS: TestIntegration/FsMkdir (23.28s)
--- PASS: TestIntegration/FsMkdir/FsEncoding (21.11s)
--- PASS: TestIntegration/FsMkdir/FsEncoding/dot (4.96s)
--- PASS: TestIntegration/FsMkdir/FsEncoding/dot_dot (5.33s)
--- PASS: TestIntegration/FsMkdir/FsEncoding/leading_dot (5.24s)
--- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_dot (5.21s)
PASS

**Comment by ncw** [comment]
Hey @kingston125 that is looking better. Can you get rid of all the encoding failures next?

Thanks

Nick

**Comment by kingston125** [comment]
> Hey @kingston125 that is looking better. Can you get rid of all the encoding failures next?
> 
> Thanks
> 
> Nick

Thank you for the quick reply. Yes, I’m trying to work on the encoding, but there’s one thing I don’t understand (previous comment) can you explain why?

go test -v -run 'TestIntegration/FsMkdir/FsEncoding/' ./backend/filelu -verbose 
it will all FAIL,
But then I run each FAIL one by one then it shows PASS e.g:
go test -v -run 'TestIntegration/FsMkdir/FsEncoding/dot' ./backend/filelu -verbose

**Comment by ncw** [comment]
Great work - that is looking better much :-)

> > So maybe the delete with the special character isn't working?
> > Most of the encoding are PASS only the punctuation FAIL, is it possible to skip the punctuation test?
> `--- FAIL: TestIntegration/FsMkdir/FsEncoding/punctuation (0.06s)`

The punctuation rclone tries is -

    !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~

If the punctuation test is failing it means either

- you need to encode some more characters
- There is a bug with url encoding (Affects `?` `&` `+` but it could be most of the characters)

Try to narrow down which punctuation character is failing (there could be more than one) by cutting down the list in the test. Once you have narrowed the failing characters down then see if you can create files with that using the web interface. If you can then it is a URL encoding problem. If you can't then add it to the encoding.

**Comment by kingston125** [comment]
> * you need to encode some more characters
> * There is a bug with url encoding (Affects `?` `&` `+` but it could be most of the characters)
> 
> Try to narrow down which punctuation character is failing (there could be more than one) by cutting down the list in the test. Once you have narrowed the failing characters down then see if you can create files with that using the web interface. If you can then it is a URL encoding problem. If you can't then add it to the encoding.

Thank you so much for your comment, the information you provided was very helpful. With your help, the tests are finally passing.
